### PR TITLE
Fix segfault in get_ao_compression_ratio

### PR DESCRIPTION
**Pull request by whitehawk**
Fix segfault in get_ao_compression_ratio

Problem:
Segfault happened in the queries with 'get_ao_compression_ratio' function in the 
target list and the argument for this function that was taken from a table. The 
problem was reproduced in the release build. In the development build the same 
queries led to assertion. 

Cause:
Function 'get_ao_compression_ratio' executes a query on 
'pg_aoseg.pg_aoseg_<OID>' table to calculate the compression ratio. A function 
that accesses hash-distributed or randomly distributed tables must be executed 
on coordinator. But the planner created plans with the 
'get_ao_compression_ratio' function which was executed on segments. Inside 
'get_ao_compression_ratio' there was an assert to check this case, but on the 
release build the assert was removed. So, the code of the function was executed 
on a segment and got unexpected results from 'pg_aoseg.pg_aoseg_', that led to a 
segfault. 

Fix:
Assert is changed to a runtime check. Now, if the planner creates a plan with 
the function which is executed on segments, the user will get a proper error 
message. It is also done for the similar functions 'gp_update_ao_master_stats' 
and 'get_ao_distribution'. Plus, unreachable code is removed from internal 
functions.

Note: it seems that a fully proper way would be to mark these functions as
PROEXECLOCATION_MASTER instead of PROEXECLOCATION_ANY, but that would require
changing of the system catalog, and it is acceptable only between major versions
update. So this approach is not used.

Note2: for an empty column-oriented table 'get_ao_compression_ratio' returns -1,
but for an empty row-oriented table 'get_ao_compression_ratio' returns 1. This
behavior is inconsistent, but it is not altered by this commit as not related
to the original issue. It has a reflection in the added test case. Be aware.

### CODE REVIEW

**Comment by BenderArenadata** [comment]
Failed job Build for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1217016

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/67751

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1217028

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1217029

**Comment by BenderArenadata** [comment]
Failed job Build for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1220033

**Comment by BenderArenadata** [comment]
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1217020

**Comment by BenderArenadata** [comment]
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1217021

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1217022

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1217023

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/67752

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1220049

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1220050

File: src/backend/access/aocs/aocssegfiles.c [hunk 1]
```diff
@@ -1708,6 +1708,8 @@ aocol_compression_ratio_internal(Relation parentrel)
 
 	Oid			segrelid = InvalidOid;
 
+	Assert(Gp_role == GP_ROLE_DISPATCH);
```
Comment (red1452) [inline src/backend/access/aocs/aocssegfiles.c]: I think this is useless Assert, because this function is called from one place, where there is the same Assert.

File: src/test/regress/output/appendonly.source [hunk 2]
```diff
@@ -1776,3 +1776,44 @@ LINE 1: explain select cmax, * from test_reporting_r;
 drop table test_reporting;
 drop table test_reporting_r;
 drop table test_reporting_c;
+-- Test AO functions with execution location on dispatcher.
+drop table if exists t1;
+drop table if exists t2;
+drop table if exists d;
```
Comment (red1452) [inline src/test/regress/output/appendonly.source]: If you use "if exists", it is better to add this lines to the "-- start_ignore" - "-- end_ignore" block.

File: src/test/regress/output/appendonly.source [hunk 3]
```diff
@@ -1776,3 +1776,44 @@ LINE 1: explain select cmax, * from test_reporting_r;
...
+create table t1 (a int) with (appendoptimized=true, orientation=row);
+create table t2 (a int) with (appendoptimized=true, orientation=column);
+create table d (tname text);
```
Comment (red1452) [inline src/test/regress/output/appendonly.source]: Maybe, rename this table to "tables_list".

File: src/backend/access/aocs/aocssegfiles.c (hunk 1 above)
Comment (whitehawk) [reply src/backend/access/aocs/aocssegfiles.c]: Ok, I placed it just in case once somebody may call it internally directly, but I guess it has a low probability. Then we'll need to remove the same assert from `aorow_compression_ratio_internal`

File: src/test/regress/output/appendonly.source (hunk 2 above)
Comment (whitehawk) [reply src/test/regress/output/appendonly.source]: fixed

File: src/test/regress/output/appendonly.source (hunk 3 above)
Comment (whitehawk) [reply src/test/regress/output/appendonly.source]: fixed

File: src/backend/access/aocs/aocssegfiles.c (hunk 1 above)
Comment (whitehawk) [reply src/backend/access/aocs/aocssegfiles.c]: done

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/68117

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1233158

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1233159

**Comment by red1452** [comment]
Remarks at the description:
```
the argument for this function taken -> the argument for this function is taken
plans with 'get_ao_compression_ratio' execution -> plans with the 'get_ao_compression_ratio' function which is executed
the function execution on a segment -> the function which is executed on segments
```

**Comment by whitehawk** [comment]
> Remarks at the description:
> 
> ```
> the argument for this function taken -> the argument for this function is taken
> plans with 'get_ao_compression_ratio' execution -> plans with the 'get_ao_compression_ratio' function which is executed
> the function execution on a segment -> the function which is executed on segments
> ```

updated the description

**Review by red1452 — DISMISSED** [review]

File: src/backend/access/appendonly/aosegfiles.c [hunk 4]
```diff
@@ -49,6 +49,13 @@
 #include "utils/numeric.h"
 #include "utils/visibility_summary.h"
 
+#define VALIDATE_GP_ROLE() \
+	if (Gp_role != GP_ROLE_DISPATCH) \
+		ereport(ERROR,\
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),\
+				 errmsg("Function '%s' can run only on dispatcher.",\
+						PG_FUNCNAME_MACRO)));
```
Comment (bandetto) [inline src/backend/access/appendonly/aosegfiles.c]: Error messages usually start with a lowercase letter and don't have a period at the end. Maybe reword this message in a better way? For example:
```suggestion
#define VALIDATE_GP_ROLE() \
	if (Gp_role != GP_ROLE_DISPATCH) \
		ereport(ERROR, \
				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED), \
				 errmsg("function '%s' must be executed on the dispatcher", \
						PG_FUNCNAME_MACRO)));
```

Comment (whitehawk) [reply src/backend/access/appendonly/aosegfiles.c]: Ok, fixed

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/68661

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1261434

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1261435

**Review by bandetto — DISMISSED** [review]

**Review by red1452 — DISMISSED** [review]

File: src/backend/access/appendonly/aosegfiles.c [hunk 5]
```diff
@@ -49,6 +49,13 @@
 #include "utils/numeric.h"
 #include "utils/visibility_summary.h"
 
+#define VALIDATE_GP_ROLE() \
+	if (Gp_role != GP_ROLE_DISPATCH) \
+		ereport(ERROR,\
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),\
+				 errmsg("function '%s' must be executed on the dispatcher",\
+						PG_FUNCNAME_MACRO)));
```
Comment (red1452) [inline src/backend/access/appendonly/aosegfiles.c]: Add "DO WHILE" to this macro to avoid problem if we write:
```
if (something)
  VALIDATE_GP_ROLE();
else
   do_things
```

File: src/backend/access/appendonly/aosegfiles.c [hunk 6]
```diff
@@ -1506,7 +1513,7 @@ gp_update_ao_master_stats(PG_FUNCTION_ARGS)
 	int64		result;
 	Snapshot	appendOnlyMetaDataSnapshot;
 
-	Assert(Gp_role == GP_ROLE_DISPATCH);
+	VALIDATE_GP_ROLE();
```
Comment (red1452) [inline src/backend/access/appendonly/aosegfiles.c:1519]: `;` is extra for macro.

File: src/test/regress/input/appendonly.source [hunk 7]
```diff
@@ -845,3 +845,34 @@ explain select cmax, * from test_reporting_r;
...
+drop table if exists t1;
+drop table if exists t2;
+drop table if exists tables_list;
+-- end_ignore
+
+create table t1 (a int) with (appendoptimized=true, orientation=row);
+create table t2 (a int) with (appendoptimized=true, orientation=column);
+create table tables_list (tname text);
+insert into t1 values (0);
+insert into t2 values (0);
```
Comment (red1452) [inline src/test/regress/input/appendonly.source]: Adding rows to `t1` and `t2` is extra. Error messages are printed without it.

Comment (whitehawk) [reply src/test/regress/input/appendonly.source]: > Adding rows to t1 and t2 is extra. Error messages are printed without it.

Ok.

But it uncovered another issue. 
For an empty column-oriented table `get_ao_compression_ratio` returns -1. I consider that an expected results, as according to function specification, it:
```
Calculates the compression ratio for a compressed append-optimized table. 
If information is not available, this function returns a value of -1.
```
But for an empty row-oriented table `get_ao_compression_ratio` returns 1. It comes from the code in `aorow_compression_ratio_internal`, which is some fix for an old compatibility issue with GPDB 3.3 tables:
```
			...
			/*
			 * Get totals from QE's and calculate the compression ratio. In
			 * tables upgraded from GPDB 3.3 the eofuncompressed column could
			 * contain NULL, this is fixed in more recent upgrades.
			 */
			char	   *attr1 = SPI_getvalue(tuple, tupdesc, 1);
			char	   *attr2 = SPI_getvalue(tuple, tupdesc, 2);

			if (NULL == attr1 || NULL == attr2)
			{
				compress_ratio = 1;
			}
			else if (scanint8(attr1, true, &eof) &&
					 scanint8(attr2, true, &eof_uncomp))
			...
```

I think that for an empty table the output of `get_ao_compression_ratio` should be consistent for both column- and row-oriented tables.
So I've updated this part of function as well.

File: src/backend/access/appendonly/aosegfiles.c (hunk 5 above)
Comment (whitehawk) [reply src/backend/access/appendonly/aosegfiles.c]: > Add "DO WHILE" to this macro to avoid problem if we write

Ok.

That could have been avoided if all single statement bodies of conditional/loop statements were wrapped in braces.
It is a mandatory requirement for many mission critical systems (for ex. refer to https://pvs-studio.com/en/docs/warnings/v2507/ ).
Unfortunately, it is not commonly used in postgres code :(.

File: src/backend/access/appendonly/aosegfiles.c (hunk 6 above)
Comment (whitehawk) [reply src/backend/access/appendonly/aosegfiles.c:1519]: > ; is extra for macro.

I would prefer to remove the `;` in the macro at the end of the loop statement. That would make the code, where the macro is used, more consistent.

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/68912

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1276287

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1276288

**Comment by BenderArenadata** [comment]
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1276279

**Comment by BenderArenadata** [comment]
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1276280

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1276281

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1276282

File: src/test/regress/input/appendonly.source (hunk 7 above)
Comment (whitehawk) [reply src/test/regress/input/appendonly.source]: Decided to discard the changes in `get_ao_compression_ratio` for empty row-oriented tables as they are not directly related to the current issue and they affect another test.

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/68925

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1277025

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1277026

**Review by red1452 — DISMISSED** [review]

File: src/backend/access/appendonly/aosegfiles.c [hunk 8]
```diff
@@ -49,6 +49,16 @@
 #include "utils/numeric.h"
 #include "utils/visibility_summary.h"
 
+#define VALIDATE_GP_ROLE() \
+	do \
+	{ \
+		if (Gp_role != GP_ROLE_DISPATCH) \
+			ereport(ERROR,\
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),\
+				 	errmsg("function '%s' must be executed on the dispatcher",\
```
Comment (bandetto) [inline src/backend/access/appendonly/aosegfiles.c]:
```suggestion
			ereport(ERROR, \
					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED), \
				 	 errmsg("function '%s' must be executed on the dispatcher", \
```

File: src/test/regress/output/appendonly.source [hunk 9]
```diff
@@ -1776,3 +1776,39 @@ LINE 1: explain select cmax, * from test_reporting_r;
...
+           ->  Seq Scan on tables_list
+ Optimizer: Postgres query optimizer
+(5 rows)
+
+select get_ao_compression_ratio(tname) from
+	unnest(array(select tname from tables_list)) as tname;
+ get_ao_compression_ratio 
+--------------------------
+                        1
+                       -1
```
Comment (bandetto) [inline src/test/regress/output/appendonly.source:1809]: Can you add explanation of this behavior to your note in the PR description? I feel like someone in the future could bump into this issue and try to look at this patch.

Comment (whitehawk) [reply src/test/regress/output/appendonly.source:1809]: Ok, added

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/69195

**Review by bandetto — APPROVED** [review]

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1290381

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1290382

**Review by red1452 — APPROVED** [review]

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/69424

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1304004

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1304006

**Comment by BenderArenadata** [comment]
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1304001
